## Business Analytics - Programming Task 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 6 (4 April 2025) @ 11.59pm  
**Submission**: Provide your answers in this Jupypter notebook and submit it via iLearn link  



## **Predictive Analysis of Credit Card Defaults**

### Objective: 
- This assignment focuses on a dataset of customers' default payments.
- The primary goal is to predict which credit card clients are likely to default using various data mining methods.

### Background: 
Traditional risk management models classify clients as either credible or not credible based on their likelihood of default. This project aims to refine this classification by identifying specific individuals who are likely to default, enhancing the precision of credit risk assessments.

Target variable
- default.payment.next.month: Default payment (1=yes, 0=no)

The dataset contains the following features

1. ID: ID of each client
2. LIMIT_BAL: Amount of given credit in dollars (includes individual and family/supplementary credit
3. SEX: Gender (1=male, 2=female)
4. EDUCATION: (1=graduate school, 2=university, 3=high school, 4=others, 5=unknown, 6=unknown)
5. MARRIAGE: Marital status (1=married, 2=single, 3=others)
6. AGE: Age in years
7. PAY_0: Repayment status in September (-1=pay duly, 1=payment delay for one month, 2=payment delay for two months, ... 8=payment delay for eight months, 9=payment delay for nine months and above)
8. PAY_2: Repayment status in August (scale same as above)
9. PAY_3: Repayment status in July, (scale same as above)
10. PAY_4: Repayment status in June (scale same as above)
11. PAY_5: Repayment status in May (scale same as above)
12. PAY_6: Repayment status in April (scale same as above)
13. BILL_AMT1: Amount of bill statement in September (dollars)  
14. BILL_AMT2: Amount of bill statement in August (dollars)  
15. BILL_AMT3: Amount of bill statement in July (dollars)  
16. BILL_AMT4: Amount of bill statement in June (dollars)  
17. BILL_AMT5: Amount of bill statement in May (dollars)  
18. BILL_AMT6: Amount of bill statement in April (dollars)   
19. PAY_AMT1: Amount of previous payment in September (dollars)  
20. PAY_AMT2: Amount of previous payment in August (dollars)  
21. PAY_AMT3: Amount of previous payment in July (dollars)   
22. PAY_AMT4: Amount of previous payment in June (dollars)  
23. PAY_AMT5: Amount of previous payment in May (dollars)   
24. PAY_AMT6: Amount of previous payment in April (dollars)  



---

**Problem 1** - Reading the dataset (Total Marks: 20)



**Q1**. Create a pandas dataframe contining the first 10,000 rows from the credit card dataset provided in the **assignment_data** folder 
- Delete 'ID' column
- Print `info()` of the dataframe 

(5 marks) 




In [5]:
# ---- provide your code here -----
import pandas as pd
# 10,000 rows of the dataset
df = pd.read_excel("assignment_data/credit_data.xlsx", nrows=10000)

# Drop ID 
df = df.drop(columns=["ID"])

# Display the structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   LIMIT_BAL                   10000 non-null  int64  
 1   SEX                         9617 non-null   float64
 2   EDUCATION                   9617 non-null   float64
 3   MARRIAGE                    9617 non-null   float64
 4   AGE                         9617 non-null   float64
 5   PAY_0                       9617 non-null   float64
 6   PAY_2                       9617 non-null   float64
 7   PAY_3                       9617 non-null   float64
 8   PAY_4                       9617 non-null   float64
 9   PAY_5                       9617 non-null   float64
 10  PAY_6                       9642 non-null   float64
 11  BILL_AMT1                   9642 non-null   float64
 12  BILL_AMT2                   9642 non-null   float64
 13  BILL_AMT3                   9642

<hr style="width:25%;margin-left:0;"> 

**Q2**. List which **features** are *numeric*, *ordinal*, and *nominal* variables, and how many features of each kind there are in the dataset.
To answer this question 

- Find the definitions of numeric, ordinal and nominal variables in the course material    
- Carefully consider what values each feature can take as well as the output of `df.info()`. 

Your answer should be written up in Markdown and include:
1) A table listing all the features present in the dataset and their type (fill out the table template provided below) and
2) A brief description of the contents of the table.

|Variable Kind|Number of Features|Feature Names
| --- | --- | --- |
| Numeric | some number | some text |
| some text  | some number | some text |
| some text  | some number | some text |


(10 marks)

The dataset consists of 24 features (after dropping the 'ID' column), which can be categorized based on their data type and meaning into numeric, ordinal, or nominal variables.

| Variable Kind | Number of Features | Feature Names |
|---------------|--------------------|----------------|
| Numeric       | 14                 | LIMIT_BAL, AGE, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6 |
| Ordinal       | 7                  | EDUCATION, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6 |
| Nominal       | 3                  | SEX, MARRIAGE, default payment next month |

### Explanation:

Variables were grouped based on how they are used and what they represent:

- **Numeric**: These are actual numbers like credit limit, age, bill amounts, and past payments. They can be measured and used in calculations.

- **Ordinal**: These are categories with order. For example, education level and repayment status have a clear ranking, even if the gaps between values aren't equal.

- **Nominal**: These are labels with no order, like gender and default payment, marriage is also part of nominal because there is no order in marital status.

This classification helps us understand how to handle each type during analysis.

<hr style="width:25%;margin-left:0;"> 

**Q3.** Missing Values. 

- Print out the number of missing values for each variable in the dataset and comment on your findings.

(5 marks)

In [10]:
# ---- provide your code here -----
# Print the number of missing values
missing_values = df.isnull().sum()
print(missing_values)

LIMIT_BAL                       0
SEX                           383
EDUCATION                     383
MARRIAGE                      383
AGE                           383
PAY_0                         383
PAY_2                         383
PAY_3                         383
PAY_4                         383
PAY_5                         383
PAY_6                         358
BILL_AMT1                     358
BILL_AMT2                     358
BILL_AMT3                     358
BILL_AMT4                     358
BILL_AMT5                     368
BILL_AMT6                     368
PAY_AMT1                      368
PAY_AMT2                      368
PAY_AMT3                       10
PAY_AMT4                       10
PAY_AMT5                      290
PAY_AMT6                      290
default payment next month      0
dtype: int64


---- provide your text answer here ----

The result shows that some variables like SEX, EDUCATION, MARRIAGE, AGE, and PAY_0 to PAY_5 have 383 missing values. Other features like BILL_AMT and PAY_AMT also have hundreds of missing values, while the target variable has no missing.

This condition shows tha**handling missing value* is needed before doing modeling. But in some cases, dropping rows or columns can also be an option, depending on how many missing and how important the feature is.

In conclusion, imputation is the main strategy here because many important features have a significant amount of missing dal.

We will handle missing values using:
- **Mean** for numeric features
- **Mode** for categorical and ordinal features

This ensures the dataset remains complete without dropping any rows.

---
---

**Problem 2.** Cleaning data and dealing with categorical features (Total Marks: 40)


**Q1.** 

- Use an appropriate `pandas` function to impute missing values using one of the following two strategies: `mean` and `mode`. (10 marks)
    - Take into consideration the type of each variable (as in Q2 above) and the best practices we discussed in class/lecture notes
- Explain what data imputation is, how you have done it here, and what decisions you had to make. (5 marks)


(Total: 15 marks)

In [14]:
# ---- provide your code here -----
# Use mean for numeric variables
numeric_cols = ['LIMIT_BAL', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
                'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6',
                'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3',
                'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6','AGE']

# Use mode for categorical and ordinal variables
categorical_cols = ['SEX', 'EDUCATION', 'MARRIAGE',
                    'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']

# Impute numeric columns with mean
for col in numeric_cols:
    df[col] = df[col].fillna(df[col].mean())

# Impute categorical/ordinal columns with mode
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

---- provide your text answer here ----

Missing value imputation is an important step to ensure the dataset is complete and ready for analysis. In this task, I used two imputation methods: mean and mode, selected based on the type of each variable.

- **Mean** was used for numerical variables such as LIMIT_BAL, AGE, BILL_AMT, and PAY_AMT.  
  This method is suitable for continuous data and helps maintain the overall distribution.

- **Mode** was applied to categorical and ordinal variables like SEX, EDUCATION, MARRIAGE, and PAY_.  
  Using the most frequent value ensures that categorical integrity is preserved.

These decisions were made by considering:
- The nature of each variable
- Best practices discussed during lectures

With these imputations, the dataset is now ready for modeling without losing important patterns in the data.

<hr style="width:25%;margin-left:0;"> 

**Q2**. 
- Print `value_counts()` of the 'SEX' column and add a dummy variable named 'SEX_FEMALE' to `df` using `get_dummies()` (3 marks)
- Carefully explain what the values of the new variable 'SEX_FEMALE' mean (2 mark)
- Make sure the variable 'SEX' is deleted from the original dataframe   

(Total: 5 marks)  

In [17]:
# ---- provide your code here -----
# Print value counts of the SEX column
sex_counts = df['SEX'].value_counts()
print(df['SEX'].value_counts())

# Create dummy variable for SEX and rename column
df = pd.get_dummies(df, columns=['SEX'], drop_first=True)

df.rename(columns={'SEX_2.0': 'SEX_FEMALE'}, inplace=True)


SEX
2.0    6162
1.0    3838
Name: count, dtype: int64


---- provide your text answer here ----

Based on the value_counts() output, the SEX column contains two values:
- **2.0** with **6,162** records
- **1.0** with **3,838** records

To prepare this categorical variable for modeling, I created a dummy variable using get_dummies() with drop_first=True. This created one dummy column, originally named SEX_2.0, which I renamed to SEX_FEMALE for clarity.

**Meaning of `SEX_FEMALE`:**
- A value of **1** indicates the person is **female (SEX = 2)**
- A value of **0** indicates the person is **male (SEX = 1)**

After creating the dummy variable, the original SEX column was automatically removed from the dataframe.  
This transformation ensures the data is in a numerical format suitable for analysis.

<hr style="width:25%;margin-left:0;"> 

**Q3**. Print `value_counts()` of the 'MARRIAGE' column and *carefully* comment on what you notice in relation to the definition of this variable. 

(Total: 5 marks) 

In [20]:
# ---- provide your code here -----
df['MARRIAGE'].value_counts()

MARRIAGE
2.0    5518
1.0    4380
3.0      82
0.0      20
Name: count, dtype: int64

---- provide your text answer here ----

The value_counts() output for the **MARRIAGE** column shows the following distribution:

- **1.0** → 5,518 records (**married**)  
- **2.0** → 4,380 records (**single**)  
- **3.0** → 82 records (**others**)  
- **0.0** → 20 records (**invalid or undefined**)

According to the dataset documentation, the valid values for **MARRIAGE** are:
- **1 = Married**
- **2 = Single**
- **3 = Others**

The presence of **0.0** indicates a data entry error or an undefined category that does not match the official label definitions. Although it appears in only a small number of cases, this value should be addressed to avoid issues during preprocessing.

One reasonable approach would be to group **0.0** under the **"Others" (3.0)** category, in order to maintain consistency in the categorical structure.


<hr style="width:25%;margin-left:0;"> 

**Q4**. 

- Apply `get_dummies()` to 'MARRIAGE' feature and add dummy variables 'MARRIAGE_MARRIED', 'MARRIAGE_SINGLE', 'MARRIAGE_OTHER' to `df`. (5 marks)   
- *Carefully consider* how to allocate all the values of 'MARRIAGE' across these 3 newly created features (5 marks)
    - Do not delete observations 
    - Do not assume that the anomaly are missing observations      
    - Explain what decision you made
- Make sure that 'MARRIAGE' is deleted from `df`   

(Total: 10 marks)   

In [23]:
# ---- provide your code here -----
# Group undefined value 0 into category 3 (Others)
df['MARRIAGE'] = df['MARRIAGE'].replace(0, 3)

# Create dummy variables for MARRIAGE
df = pd.get_dummies(df, columns=['MARRIAGE'], prefix='MARRIAGE')

# Rename dummy variable y
df.rename(columns={
    'MARRIAGE_1.0': 'MARRIAGE_MARRIED',
    'MARRIAGE_2.0': 'MARRIAGE_SINGLE',
    'MARRIAGE_3.0': 'MARRIAGE_OTHER'
}, inplace=True)

---- provide your text answer here ----

Applied get_dummies() to the MARRIAGE feature to convert it into three dummy variables: **MARRIAGE_MARRIED**, **MARRIAGE_SINGLE**, and **MARRIAGE_OTHER**.

Before creating the dummies, I identified that the **MARRIAGE** column contained a value **0.0**, which is not part of the official categories (1 = Married, 2 = Single, 3 = Others).  
As instructed, I did not treat this value as missing, nor delete any observations. I decided to **reassign the value 0.0 into category 3 (Others)**, as it is undefined and appears in a very small portion of the data.

After using get_dummies(), I renamed the dummy columns for clarity:
- MARRIAGE_1.0 → **MARRIAGE_MARRIED**
- MARRIAGE_2.0 → **MARRIAGE_SINGLE**
- MARRIAGE_3.0 → **MARRIAGE_OTHER**

Each dummy variable takes the value:
- 1 if the observation belongs to that category
- 0 otherwise

The original **MARRIAGE** column was automatically removed by the get_dummies() function, and the data is now ready for further analysis or modeling.

<hr style="width:25%;margin-left:0;"> 

**Q5**. In the column 'EDUCATION', convert the values {0, 5, 6} to the value 4. 

(Total: 5 marks)  

In [26]:
# ---- provide your code here -----
# Replace 0, 5, and 6 in EDUCATION with 4 (Others)
df['EDUCATION'] = df['EDUCATION'].replace([0, 5, 6], 4)

---
---

**Problem 3** Preparing X and y arrays (Total Marks: 10)

**Q1**. 

- Create a numpy array `y` from the first 7,000 observations of `default payment next month` column from `df` (2.5 marks)   
- Create a numpy array `X`  from the first 7,000 observations of all the remaining variables in `df` (2.5 marks)   

(Total: 5 Marks)

In [29]:
# ---- provide your code here -----

# Create X and y arrays from the first 7,000 observations
X = df.drop(columns=['default payment next month']).iloc[:7000]
y = df['default payment next month'].iloc[:7000]

<hr style="width:25%;margin-left:0;"> 

**Q2**. 

- Use an appropriate `sklearn` library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 31 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data to mean zero and variance one using an approapriate `sklearn` library (2.5 marks)   


(Total: 5 marks) 

In [31]:
# ---- provide your code here -----

from sklearn.model_selection import train_test_split

# Split the data 70% train and 30% test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=31,stratify=y)

from sklearn.preprocessing import StandardScaler

# Standardise the features using StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)  # Fit on train, transform train
X_test = scaler.transform(X_test)        # Transform test with same scaler

---
---

**Problem 4**. Training Models and Interpretation (Total Marks: 30)


**Q1**. 

- Train one linear classifier we studied in class using standardised data (6 marks)
- Compute and print training and test dataset accuracies (4 marks)

(Total: 10 marks)   

In [34]:
# ---- provide your code here -----

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Train Logistic Regression model
model = LogisticRegression(max_iter=1000, random_state=31)

# Train model
model.fit(X_train, y_train)

# Make prediction 
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Calculate Accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)

# Print Accuracy
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")

Training Accuracy: 82.06%
Test Accuracy: 80.52%


<hr style="width:25%;margin-left:0;"> 

**Q2.**

- Train one nonlinear classifier we studied in class on the same dataset (6 marks)
- Compute and print training and test dataset accuracies (4 marks)


(Total: 10 marks)  

In [36]:
# ---- provide answer here -----

from sklearn.ensemble import RandomForestClassifier

# Train Random Forest model
rf_model = RandomForestClassifier(n_estimators=100, random_state=31)
rf_model.fit(X_train, y_train)

# Compute accuracies
rf_train_accuracy = rf_model.score(X_train, y_train)
rf_test_accuracy = rf_model.score(X_test, y_test)

print(f"Training Accuracy: {rf_train_accuracy * 100:.2f}%")
print(f"Test Accuracy: {rf_test_accuracy * 100:.2f}%")

Training Accuracy: 99.98%
Test Accuracy: 80.71%


<hr style="width:25%;margin-left:0;"> 

**Q3**. 

- Comment on the accuracy results obtained from the two classifiers (6 marks)
- Based on our investigation into credit card default predictions in this assignment, which model would you recommend? (4 marks)


(Total: 10 marks)     


---- provide your text answer here ----

**Comment on the Accuracy Results**

In this task, I trained two types of classifiers based on the approaches studied in class:

- **Linear classifier**: Logistic Regression  
  - Training Accuracy: 82.06%  
  - Testing Accuracy: 80.52%

- **Nonlinear classifier**: Random Forest  
  - Training Accuracy: 99.98%  
  - Testing Accuracy: 80.71%

The results show that the **nonlinear classifier (Random Forest)** achieved slightly higher test accuracy than the linear classifier. However, the large gap between training and testing accuracy indicates **overfitting**, meaning the model fits the training data too closely and may not generalize well.

In contrast, the **linear classifier (Logistic Regression)** shows more balanced results between training and testing accuracy, suggesting **better generalization** to unseen data.

---

**Model Recommendation**

Although Logistic Regression is more stable and interpretable, I recommend using the **nonlinear classifier (Random Forest)** because:

- It achieves the **highest test accuracy**  
- It can model **complex relationships** between features  
- It is suitable for this dataset, which includes both numerical and categorical features

**Conclusion:**  
The nonlinear classifier is more suitable for this task due to its higher predictive performance. However, attention should be given to the risk of overfitting when applying this model in practice.

---
---

## Marking Criteria

To achieve a perfect score, your solutions must adhere to the criteria outlined below:

- Ensure that all numerical answers are accurate.
- Utilize the exact Python functions and libraries specified within the assignment instructions.
- For any written responses, provide accurate information, articulated in clear, complete sentences.
- Do not add extra cells beyond what is provided in the notebook.
- Do not print output with your code unless explicitly instructed to do so.
- Maintain a clean and organised notebook layout that is easy to follow.
- Marks will be deducted for not following the above instructions.
    
---
---